filebeat/beater/filebeat.go

```diff
@@ -108,22 +113,42 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	defer func() {
 		// With harvesters being stopped, optionally wait for all enqueued events being
 		// published and written by registrar before continuing shutdown.
-		fb.sigWait.Wait()
+		if !*once {
+			fb.sigWait.Wait()
+		}
 
-		// continue shutdown
+		// Closes publisher so no further events can be sent
 		publisherChan.Close()
+		// Stopping spooler
 		spooler.Stop()
 	}()
 
-	err = crawler.Start(registrar.GetStates())
+	err = crawler.Start(registrar.GetStates(), *once)
 	if err != nil {
 		return err
 	}
 	// Stop crawler -> stop prospectors -> stop harvesters
 	defer crawler.Stop()
 
-	// Blocks progressing. As soon as channel is closed, all defer statements come into play
-	<-fb.done
+	// Wait for completion if running once.
+	// Running the shutdown in a go routing is required for CTRL-C to still work for run-once
+
+	if *once {
+		runOnce := func() {
+			logp.Info("Running filebeat once. Waiting for completion ...")
+			crawler.WaitForCompletion()
+
+			// Wait for registrar to finish writing registry
+			wgEvents.Wait()
```

Comment:
don't wait here. just add `wgEvents.Wait` to sigWait after waiting for shutdown:

```
...
fb.sigWait.Wait()

waitPublished := fb.config.ShutdownTimeout > 0 || *once
if waitPublished {
   fb.sigWait.Add(wgEvents.Wait)
   if fb.config.ShutdownTimeout > 0 {
    fb.sigWait.AddTimeout(fb.config.ShutdownTimeout)
  } 
}

```